File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        int i = 0, j = 0, mxlen = 0, cnt = 0;
        unordered_map<char, int> mp;
        while(j < n){
            mp[s[j]]++;
           cnt++;
            while(mp[s[j]] > 1){
                mxlen = max(mxlen, cnt-1);
                mp[s[i]]--;
                cnt = j-i;
                i++;
            }
            j++;
        }
        return max(cnt,mxlen);
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character > st  = new HashSet<>();

        int i = 0, j = 0, maxLen = 0;
        while(j < s.length()){
            char ch = s.charAt(j);
            while(st.contains(ch)){
                st.remove(s.charAt(i));
                i++;
            }
            maxLen = Math.max(maxLen, j-i+1);
            st.add(ch);
            j++;
        }
        return maxLen;
    }
}